pyccel/ast/datatypes.py

```diff
@@ -138,215 +224,449 @@ def __reduce__(self):
         """
         return (self.__class__, ())
 
-class NativeBool(DataType, metaclass=Singleton):
+    def switch_basic_type(self, new_type):
+        """
+        Change the basic type to the new type.
+
+        Change the basic type to the new type. In the case of a FixedSizeType the
+        switch will replace the type completely, directly returning the new type.
+
+        Parameters
+        ----------
+        new_type : PyccelType
+            The new basic type.
+
+        Returns
+        -------
+        PyccelType
+            The new type.
+        """
+        assert isinstance(new_type, FixedSizeType)
+        return new_type
+
+class FixedSizeNumericType(FixedSizeType):
     """
-    Class representing a boolean datatype.
+    Base class representing a scalar numeric datatype.
 
-    Class representing a boolean datatype.
+    The base class representing a scalar numeric datatype which can be
+    represented in memory. E.g. int32, int64.
+    """
+    __slots__ = ()
+
+    @property
+    def precision(self):
+        """
+        Precision of the datatype of the object.
+
+        The precision of the datatype of the object. This number is related to the
+        number of bytes that the datatype takes up in memory. For basic types the
+        number is equivalent to the number of bytes in memory (e.g. `float64` has
+        precision = 8 as it takes up 8 bytes), however for less simple types the
+        connection is less trivial. For example `complex128` has precision = 8 as
+        it is comprised of two `float64` objects (which have precision=8).
+        It should be noted that this is not the convention chosen by NumPy (in NumPy
+        a `complex128` is so named because `16*8=precision*bits_in_a_byte=128`).
+
+        The precision in Pyccel is equivalent to the `kind` parameter in Fortran.
```

Comment:
What are the reasons behind following Fortran Conventions for `precision` and not following Numpy's?

Reply:
See above